Add shinyloadtest and shinytest2 vignette (#210)
rstudio/shinytest2: vignettes/use-application-audit.Rmd

---
title: "use-application-audit"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Monkey testing}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(shinytest2)
library(shiny)
library(deSolve)
```

# Audit Shiny apps with {shinytest2}

Have you ever dreamed of profiling, load testing your Shiny app at each commit, 
without having to manually run any script?

In this vignette, we'll see how one can design and automate Shiny apps 
audit pipelines with `{shinytest2}`.  

## Introduction

This case study consists of analyzing a particularly non-optimized app, whose 

> code is defined below and located in `vignettes/non-optimized-app/app.R`. 

Author: Done!

    args = args,
    stderr= "",
    stdout = ""
  )

  while (any(is.na(pingr::ping_port("127.0.0.1", 3515)))) {
    message("Waiting for Shiny app to start...")
    Sys.sleep(0.1)
  }

  attempt::stop_if_not(
    process$is_alive(),
    msg = "Unable to launch the subprocess"
  )

  process
}
```

where `r_bg()` starts a background R process, passing the corresponding function 
and parameters. Besides, we provide some log elements and safety guard in case 
the app can't start. To launch the app and recorder we simply call:

``` r
target <- start_r_bg(shiny_bg, path = "vignettes/non-optimized-app/app.R")
# Listening on 127.0.0.1:3515
recorder <- start_r_bg(recorder_bg)
# Listening on 127.0.0.1:8600
```

### Connect Chrome

The previous part was the most technical step. Now, we only have to start a 
Chrome headless browser on port `8600`, where the load test recorder runs. 
You'll notice that `{shinytest2}` also supports remote urls. 
We ensure to set a consistent load timeout, which will avoid to see this kind of 
annoying `{shinytest2}` warning:

```
{shinytest2} R  info 15:25:37.36 Error while initializing AppDriver:
Shiny app did not become stable in 10000ms.
```

``` r

> # start headless chrome (points to recorder!).
> # AppDriver also support remote urls.

Reviewer: ```suggestion
# Start AppDriver with recorder url
```

> chrome <- shinytest2::AppDriver$new("http://127.0.0.1:8600", load_timeout = 15000)

Reviewer: Spread arg calculation so it can easily be read ; Also changing name to `app`

```suggestion
chrome <- shinytest2::AppDriver$new("http://127.0.0.1:8600", load_timeout = 15 * 1000)
```

```

If your running under Linux OS, below shows the list of R processes running 
so far in the background:

```
$ netstat -lntp

Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:8600          0.0.0.0:*               LISTEN      361115/R            
tcp        0      0 127.0.0.1:3515          0.0.0.0:*               LISTEN      361101/R            
tcp        0      0 127.0.0.1:44179         0.0.0.0:*               LISTEN      359555/google-chrom
```

If everything is successful, you should see a __Client connected__ message in the 
R console. We then change the `mu` parameter input without forgetting the 
timeout:

``` r

> chrome$set_inputs(mu = 4, timeout_ = 15000)

Reviewer: ```suggestion
app$set_inputs(mu = 4, timeout_ = 15 * 1000)
```

```

We can inspect the logs with the help of `chrome$get_log()`, to check whether 
everything run smoothly:

```

> chrome$get_log()

Reviewer: ```suggestion
app$get_logs()
```

> # {shinytest2} R  info 14:27:00.80 Start AppDriver initialization
> # {shinytest2} R  info 14:27:00.82 Creating new ChromoteSession
> # {shinytest2} R  info 14:27:02.20 Navigating to Shiny app
> # {shinytest2} R  info 14:27:02.57 Injecting shiny-tracer.js
> # {chromote}   JS info 14:27:02.61 shinytest2; jQuery not found
> # {chromote}   JS info 14:27:02.63 shinytest2; Loaded
> # {shinytest2} R  info 14:27:02.64 Waiting for Shiny to become ready
> # {chromote}   JS info 14:27:02.72 shinytest2; jQuery found
> # {chromote}   JS info 14:27:02.73 shinytest2; Waiting for shiny session to connect
> # {chromote}   JS info 14:27:03.12 shinytest2; Connected
> # {shinytest2} R  info 14:27:03.17 Waiting for Shiny to become idle for 200ms within 10000ms
> # {chromote}   JS info 14:27:03.18 shinytest2; Waiting for Shiny to be stable
> # {chromote}   JS info 14:27:03.27 shinytest2; shiny:busy
> # {chromote}   JS info 14:27:06.28 shinytest2; shiny:idle
> # {chromote}   JS info 14:27:06.30 shinytest2; shiny:value brussels
> # {chromote}   JS info 14:27:06.48 shinytest2; Shiny has been idle for 200ms
> # {shinytest2} R  info 14:27:06.48 Shiny app started
> # {shinytest2} R  info 14:30:17.79 Setting inputs: 'mu'', ''X'', ''Y'
> # {chromote}   JS info 14:30:17.81 shinytest2; inputQueue: adding mu
> # {chromote}   JS info 14:30:17.81 shinytest2; inputQueue: adding X
> # {chromote}   JS info 14:30:17.81 shinytest2; inputQueue: adding Y
> # {chromote}   JS info 14:30:17.82 shinytest2; inputQueue: flushing mu
> # {chromote}   JS info 14:30:17.83 shinytest2; inputQueue: flushing X
> # {chromote}   JS info 14:30:17.83 shinytest2; inputQueue: flushing Y
> # {chromote}   JS info 14:30:17.87 shinytest2; shiny:busy
> # {chromote}   JS info 14:30:20.42 shinytest2; shiny:idle
> # {chromote}   JS info 14:30:20.46 shinytest2; shiny:value brussels
> # {shinytest2} R  info 14:30:20.47 Finished setting inputs. Timedout: FALSE

Reviewer: ```suggestion
#> {shinytest2} R  info 14:27:00.80 Start AppDriver initialization
#> {shinytest2} R  info 14:27:00.82 Creating new ChromoteSession
#> {shinytest2} R  info 14:27:02.20 Navigating to Shiny app
#> {shinytest2} R  info 14:27:02.57 Injecting shiny-tracer.js
#> {chromote}   JS info 14:27:02.61 shinytest2; jQuery not found
#> {chromote}   JS info 14:27:02.63 shinytest2; Loaded
#> {shinytest2} R  info 14:27:02.64 Waiting for Shiny to become ready
#> {chromote}   JS info 14:27:02.72 shinytest2; jQuery found
#> {chromote}   JS info 14:27:02.73 shinytest2; Waiting for shiny session to connect
#> {chromote}   JS info 14:27:03.12 shinytest2; Connected
#> {shinytest2} R  info 14:27:03.17 Waiting for Shiny to become idle for 200ms within 10000ms
#> {chromote}   JS info 14:27:03.18 shinytest2; Waiting for Shiny to be stable
#> {chromote}   JS info 14:27:03.27 shinytest2; shiny:busy
#> {chromote}   JS info 14:27:06.28 shinytest2; shiny:idle
#> {chromote}   JS info 14:27:06.30 shinytest2; shiny:value brussels
#> {chromote}   JS info 14:27:06.48 shinytest2; Shiny has been idle for 200ms
#> {shinytest2} R  info 14:27:06.48 Shiny app started
#> {shinytest2} R  info 14:30:17.79 Setting inputs: 'mu'', ''X'', ''Y'
#> {chromote}   JS info 14:30:17.81 shinytest2; inputQueue: adding mu
#> {chromote}   JS info 14:30:17.81 shinytest2; inputQueue: adding X
#> {chromote}   JS info 14:30:17.81 shinytest2; inputQueue: adding Y
#> {chromote}   JS info 14:30:17.82 shinytest2; inputQueue: flushing mu
#> {chromote}   JS info 14:30:17.83 shinytest2; inputQueue: flushing X
#> {chromote}   JS info 14:30:17.83 shinytest2; inputQueue: flushing Y
#> {chromote}   JS info 14:30:17.87 shinytest2; shiny:busy
#> {chromote}   JS info 14:30:20.42 shinytest2; shiny:idle
#> {chromote}   JS info 14:30:20.46 shinytest2; shiny:value brussels
#> {shinytest2} R  info 14:30:20.47 Finished setting inputs. Timedout: FALSE
```

```

Once satisfied, we may close the headless connection to stop the recorder:

``` r
# clean

> chrome$stop()

Reviewer: ```suggestion
app$stop()
```

```

Some error message in the `shinycannon` logs can be explained by failures during 
the `{shinytest2}` driver initialization. If this error persist, best practice 
is to restart R, cleanup everything and start again. 

### Report generation

The report generation is quite straightforward:

``` r
# Close the running app
target$kill()

# Treat data and generate report
df <- shinyloadtest::load_runs("run1")
shinyloadtest::shinyloadtest_report(
  df,
  "public/index.html",
  self_contained = TRUE,
  open_browser = FALSE
)
```

We don't forget to clean the target app so as to kill the underlying process. 
Note that if you want to deploy the report on GitHub Pages, you have to name it 
`index.html`.

### Automating with Github Actions

Now, it is time to integrate the current pipeline in a CI/CD workflow. 
For convenience, we wrap all the previous steps in a single function:

``` r
record_loadtest <- function(path, timeout = 15, workers = 5) {
  message("\n---- BEGIN LOAD-TEST ---- \n")
  # start app + recorder
  target <- start_r_bg(shiny_bg, path = path)
  recorder <- start_r_bg(recorder_bg)

  # start headless chrome (points to recorder!).
  # AppDriver also support remote urls.

>   chrome <- shinytest2::AppDriver$new(

Reviewer: ```suggestion
  app <- shinytest2::AppDriver$new(
```

    "http://127.0.0.1:8600",
    load_timeout = timeout * 1000
  )

>   chrome$set_inputs(mu = 4, timeout_ = timeout * 1000)

Reviewer: ```suggestion
  app$set_inputs(mu = 4, timeout_ = timeout * 1000)
```

  # clean

>   chrome$stop()

Reviewer: ```suggestion
  app$stop()
```

  # needed to avoid
  # java.lang.IllegalStateException: last event in log not a
  # WS_CLOSE (did you close the tab after recording?)
  Sys.sleep(2)

  # shinycannon (maybe expose other params later ...)
  target_url <- "http://127.0.0.1:3515"
  system(
    sprintf(
      "shinycannon recording.log %s --workers %s --loaded-duration-minutes 2 --output-dir run1",
      target_url, workers
    )
  )

  target$kill()

  # Treat data and generate report
  df <- shinyloadtest::load_runs("run1")
  shinyloadtest::shinyloadtest_report(
    df,
    "public/index.html",
    self_contained = TRUE,
    open_browser = FALSE
  )
}
```

Below is the necessary GitHub Actions yaml file. Overall, this will:

- Run each time code is pushed to GitHub.
- Run in parallel on 3 different Ubuntu flavors.

> - Install R, some system dependencies and R packages. Install Chrome and 
> shinycannon tools for the headless loadtest.

Reviewer: ```suggestion
- Install R, some system dependencies, and R packages. Install shinycannon tools for load-testing.
```

- Deploy the load test report to GitHub Pages.

> ```yaml

Reviewer: This action should be added to the repo to run to make sure it works

on:
  push:
    branches: [main, master]
  pull_request:
    branches: [main, master]

name: shiny-loadtest-ci

jobs:
  shiny-loadtest-ci:
    runs-on: ${{ matrix.config.os }}
    permissions:
      contents: write

    name: ${{ matrix.config.os }} (${{ matrix.config.r }})

    strategy:
      fail-fast: false
      matrix:
        config:
          - {os: ubuntu-latest,   r: 'devel', http-user-agent: 'release'}
          - {os: ubuntu-latest,   r: 'release'}
          - {os: ubuntu-latest,   r: 'oldrel-1'}

    env:
      GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
      R_KEEP_PKG_SOURCE: yes

    steps:
      - uses: actions/checkout@v2
      - uses: r-lib/actions/setup-r@v2
      - uses: r-lib/actions/setup-r-dependencies@v2
        with:
          cache-version: 2
          extra-packages: |
            any::shinyloadtest
            any::lubridate
            any::DT
            any::callr
            any::shinytest2
            any::deSolve
            any::attempt

>       - name: Install shinycannon 💥 and Chrome 🌐

Reviewer: ```suggestion
      - name: Install shinycannon 💥
```

        run: |
          sudo bash -c 'apt-get update; apt-get install -y default-jre-headless; apt-get clean; rm -rf /var/lib/apt/lists/*'

>           wget https://github.com/rstudio/shinycannon/releases/download/v1.1.3/shinycannon_1.1.3-dd43f6b_amd64.deb https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb

Reviewer: Chrome is handled in pak installation in prior step. Removing chrome install. Keeping shinycannon install

```suggestion
          wget https://github.com/rstudio/shinycannon/releases/download/v1.1.3/shinycannon_1.1.3-dd43f6b_amd64.deb
```

          sudo dpkg -i ./*.deb

      - name: Run load test 🏥
        shell: Rscript {0}

>         run: source("R/audit-app.R"); record_loadtest(path = "app.R");

Reviewer: ```suggestion
        run: |
            source("R/audit-app.R")
            record_loadtest(path = "app.R");
```

>         run: source("R/audit-app.R"); record_loadtest(path = "app.R");

Reviewer: We need to describe where `./R/audit-app.R`. Should this also be included in the repo?

Author: Could be included in the repo but don't know where. Maybe in `R` or `inst/audit-app`

Author: Ok I see later you suggest `file.copy(system.file("gha/audit-app.R", package = "shinytest2")`. I am fine with this.

      - name: Deploy to GitHub pages 🚀
        if: github.event_name != 'pull_request'
        uses: JamesIves/github-pages-deploy-action@4.1.4
        with:
          clean: false
          branch: gh-pages
          folder: public
```

If you want to test it on your end, below are lines of code to setup a 

> RStudio project and link it to github:

Reviewer: ```suggestion
RStudio project and link it to GitHub:
```

``` r
path <- tempdir()
usethis::create_project(path, open = FALSE)
# Inside the project
workflows_path <- file.path(path, ".github/workflows")
r_path <- file.path(path, "R")
dir.create(workflows_path, recursive = TRUE)
dir.create(r_path)
# Copy all R helpers
file.create(file.path(r_path, "audit-app.R"))
# Copy the previous yaml template inside this file
file.create(file.path(workflows_path, "shiny-loadtest-ci.yaml"))
# copy your app content

> file.create("app.R") 

Reviewer: This would be great if it could be something that could _almost_ be copy/paste-able. Something like...

```r
# Inside the project
# Use audit GitHub Actions workflow
usethis::use_github_action(url = "https://raw.githubusercontent.com/rstudio/shinytest2/main/.github/workflows/app-audit.yaml")
# Copy audit script
dir.create(file.path(path, "R"))
file.copy(system.file("gha/audit-app.R", package = "shinytest2"), file.path(path, "R/audit-app.R"))
# Create your app.R file
file.create("app.R") 

## Need to copy in your app code to `app.R`!
```

usethis::use_description()
usethis::use_git()
usethis::use_git_protocol("ssh")
usethis::use_github()
usethis::use_github_pages()
```

An example is available [here](https://divadnojnarg.github.io/RtmpXU8dBM/).

```{r echo=FALSE, eval=TRUE, out.width='100%', fig.align='center'}
knitr::include_graphics("images/shinytest2-loadtest.png")
```

As shown in the above report, especially in the session duration tab, 

> the app is clearly not able to handle 5 simultaneous user:

Reviewer: ```suggestion
the app is clearly not able to handle 5 simultaneous user due to the very large computations being performed by the Shiny server:
```
